src/main/java/io/cdap/plugin/gcp/common/GCPErrorDetailsProviderUtil.java

```diff
@@ -69,6 +70,27 @@ public static ProgramFailureException getProgramFailureException(HttpResponseExc
       ErrorCodeType.HTTP, statusCode.toString(), externalDocumentationLink, e);
   }
 
+  /**
+   * Get a ProgramFailureException with the given error
+   * information from {@link ApiException}.
+   *
+   * @param e The ApiException to get the error information from.
+   * @return A ProgramFailureException with the given error information.
+   */
+  public static ProgramFailureException getProgramFailureException(ApiException e, String externalDocUrl,
+                                                                   @Nullable ErrorContext errorContext) {
+    Integer statusCode = e.getStatusCode().getCode().getHttpStatusCode();
+    ErrorUtils.ActionErrorPair pair = ErrorUtils.getActionErrorByStatusCode(statusCode);
+    String errorReason = String.format("%s %s. %s. For more details, see %s", statusCode, e.getMessage(),
+     pair.getCorrectiveAction(), externalDocUrl);
+    String errorMessage = e.getMessage();
+    return ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN), errorReason,
+      errorContext != null ?
+        String.format(GCPErrorDetailsProvider.ERROR_MESSAGE_FORMAT, errorContext.getPhase(), e.getClass().getName(),
+          errorMessage) : String.format("%s: %s", e.getClass().getName(), errorMessage), pair.getErrorType(), true,
+      ErrorCodeType.HTTP, statusCode.toString(), externalDocUrl, e);
+  }
```

Comment:
On a second thought, since HttpResponseException has more details then ApiException and they can co-exist in causalChain when it is a rest API call and not a grpc call, can you please traverse the causalChain again in this method to check if there is a `HttpResponseException` underneath?

Reply:
or we could do something like this in `getHttpResponseExceptionDetailsFromChain()`:

```
// First pass: Check for ProgramFailureException (avoid unnecessary re-wrapping)
for (Throwable t : causalChain) {
    if (t instanceof ProgramFailureException) {
        return (ProgramFailureException) t;
    }
}

// Second pass: Reverse iterate to prioritize HttpResponseException over ApiException
for (int i = causalChain.size() - 1; i >= 0; i--) {
    Throwable t = causalChain.get(i);

    if (t instanceof HttpResponseException) {
        return getProgramFailureException((HttpResponseException) t, externalDocUrl, null);
    }
    if (t instanceof ApiException) {
        return getProgramFailureException((ApiException) t, externalDocUrl, null);
    }
}
```

Reply:
can you please do what I suggested in this comment: https://github.com/data-integrations/google-cloud/pull/1524/files#r1983581525 to make things look cleaner?